Changed usercase prefix to "user/"

kaapstorm opened the pull request:
Don't accept colons in case property names. Added doctest for validate_property. 

(Disallowing colons reverts 7a03a32f)

@czue, @snopoke for context

kaapstorm committed:
Changed usercase prefix to "user/"

Don't colons in case property names
Added doctest

snopoke commented:
Don't you also need to update `xform.py` to not split these as if they are referring to indexed cases?

kaapstorm commented:
Labelling "hold off on merge" until usercase stuff is managed separately from indexed cases.

kaapstorm committed:
Merge branch 'master' into usercase

snopoke commented:
Seems like this won't be needed at all.

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Renamed is_usercase_enabled to is_usercase_in_use

This avoids ambiguity wetween whether the toggle is on, and whether the
usercase is currently being used by the application.

kaapstorm committed:
Implement usercase as separate form actions

kaapstorm committed:
get_case_updates deals with the right form actions

For usercase properties, use the update_usercase action, and for normal cases,
use the update_case action

kaapstorm committed:
Interface changes for user case management

kaapstorm committed:
User case ID no longer needed in the session.

Just use form actions for usercase management

dimagimon commented on corehq/apps/app_manager/util.py:
```diff
@@ -6,6 +6,7 @@
 from corehq.apps.app_manager.exceptions import SuiteError
 from corehq.apps.builds.models import CommCareBuildConfig
 from corehq.apps.app_manager.tasks import create_user_cases
+from corehq.toggles import USER_AS_A_CASE
```
F401 'USER_AS_A_CASE' imported but unused

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Unused import

kaapstorm committed:
XForm changes for usercase actions

kaapstorm committed:
Missed applyBindings for #usercase-config-ko

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Fix Save button layout

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
User case save button

kaapstorm committed:
Merge branch 'master' into usercase

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -315,6 +315,9 @@ class FormActions(DocumentSchema):
     referral_preload = SchemaProperty(PreloadAction)
 
+    update_usercase = SchemaProperty(UpdateCaseAction)
+    usercase_preload = SchemaProperty(PreloadAction)
```
Would be nice to name these with the same schema e.g. usercase_preload, usercase_update.

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -1134,11 +1140,17 @@ def extended_build_validation(self, error_meta, xml_valid, validate_module=True)
             format_key = self.get_case_property_name_formatter()
+            if case_type == USERCASE_TYPE:
+                return [format_key(*item)
+                        for item in self.actions.update_usercase.update.items()]
```
This could just be an if else:
```
if self.get_module().case_type == case_type:
    ....
elif case_type == USERCASE_TYPE:
   ....
```

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -3585,8 +3597,18 @@ def validate_lang(lang):
+    Validate a case property name
+
+    >>> validate_property('parent/maternal-grandmother_fullName')
+    >>> validate_property('user:full_name')
```
I don't think this like is relevant anymore is it?

snopoke commented on corehq/apps/app_manager/static/app_manager/js/case-config-ui-2.js:
```diff
@@ -587,6 +786,53 @@ var CaseConfig = (function () {
+                update_usercase: {
+                    update: case_properties
+                },
+                case_preload: {
```
shouldn't this be `usercase_preload`

kaapstorm replied:
Well spotted!

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1502,16 +1502,25 @@ def get_auto_select_assertions(self, case_id_xpath, mode, locale_arguments=None)
+                    'case_type': USERCASE_TYPE,
+                    'requires_selection': False,
+                    'action': actions['usercase_preload']
+                })
```
Won't this end up with two datums if you preloading and updating?

kaapstorm replied:
Do I need to add one datum for each action, or can I list two actions for the same datum?

snopoke commented on corehq/apps/app_manager/util.py:
```diff
@@ -166,16 +166,13 @@ def get_properties(self, case_type, already_visited=(),
 
-        # prefix user case properties with "user:".
+        # prefix user case properties.
         prefix_user = lambda p: USERCASE_PREFIX + p if case_type == USERCASE_TYPE else p
```
Is this necessary anymore?

kaapstorm committed:
Replace if...in with if...elif

Easier to read.

kaapstorm committed:
Updated misleading doctest

kaapstorm committed:
Typo

kaapstorm committed:
USERCASE_PREFIX not used right now.

If we need it again (e.g. case list filtering) add it back later.

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Append one datum for usercase

No need to set datum action.

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
usercase UI progress

kaapstorm committed:
Add a 1em bottom margin below the save button

otherwise it touches the well below it, and looks dumb.

kaapstorm committed:
Add top and bottom layout classes

Add .top* and .bottom* classes for layout. Put
lib/bootstrap/less/utilities.less back the way it was.

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'usercase' into save_button_bottom_margin

kaapstorm committed:
Styling based on Jenny's feedback

kaapstorm committed:
Added README file for custom styles

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'usercase' into save_button_bottom_margin

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'save_button_bottom_margin' into usercase

snopoke committed:
fix case preload action

snopoke committed:
disable user as a case tab for advanced forms

snopoke committed:
fix get usercase properties

snopoke committed:
allow referencing usercase properties in caselist and casedetails

snopoke committed:
return correct type from get_usercase_properties

snopoke committed:
include usercase properties in get_call_case_properties

snopoke committed:
rename update_usercase to usercase_update

snopoke committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/util.py
	corehq/apps/app_manager/views.py

dimagimon commented on corehq/apps/app_manager/views.py:
```diff
@@ -96,10 +95,13 @@
     enable_usercase,
     actions_use_usercase,
-    get_per_type_defaults)
+    get_usercase_properties, 
```
W291 trailing whitespace

snopoke committed:
add [1] selector to user case xpath

kaapstorm commented:
Thanks for all this @snopoke!

kaapstorm committed:
is_usercase_enabled renamed to is_usercase_in_use

kaapstorm committed:
Don't import usercase toggle specifically

kaapstorm committed:
Pass case_type to get_details

kaapstorm committed:
Don't shadow variable name from outer scope

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Use get_id to get the usercase id

kaapstorm committed:
usercase_update condition set to "always"

If the condition is not set, the usercase_update action will not be active, and the case block will not be added to the xform.

kaapstorm committed:
get_all_case_properties excludes usercase

usercase properties handled separately by get_usercase_properties

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Include usercase in get_all_case_properties

kaapstorm committed:
Include usercase in get_case_types

Add test mocks where necessary

kaapstorm committed:
Add another is_usercase_in_use mock

kaapstorm committed:
Merge branch 'master' into usercase

snopoke commented on corehq/apps/app_manager/templates/app_manager/partials/usercase_config.html:
```diff
@@ -0,0 +1,151 @@
+    <p class="help-inline" data-bind="visible: !required() && !path() && key()">{% trans "Not assigned" %}</p>
+</script>
+
+<script type="text/html" id="usercase-config:case-properties:property">
```
A lot of these look like identical copies for the ones in case_config_shared.html. Would be nice if we didn't have to copy them here.

snopoke commented on corehq/apps/app_manager/views.py:
```diff
@@ -1947,8 +1956,8 @@ def edit_form_actions(request, domain, app_id, module_id, form_id):
-    if actions_use_usercase(form.actions) and not is_usercase_enabled(domain):
-        if toggle_enabled_shortcut('user_as_a_case', domain, namespace='domain'):
+    if actions_use_usercase(form.actions) and not is_usercase_in_use(domain):
+        if toggles.USER_AS_A_CASE.enabled(domain):
```
This is a bit confusing. The interaction between the toggle and the domain setting is quite weird.

kaapstorm replied:
Ha! I noticed that last night. Not only is it confusing, it's not really necessary. Taking it out.

snopoke commented:
Merging now but need to address a few issues.

kaapstorm commented:
Thanks Simon. I'll get onto it.
